### DISCUSSION TITLE
Direct3D templates

### DISCUSSION THREAD

**Discussion by BMarques** [amerkoleci/Vortice.Windows#348] [Ideas]
To make it easier to start with Vortice Windows, I was thinking that it would be nice to have some sort of ready made visual studio templates, such as https://github.com/walbourn/directx-vs-templates. Does it make sense?

I already am working on one for myself with Direct3D 11 based on the [d3d11game_win32_dr](https://github.com/walbourn/directx-vs-templates/tree/main/d3d11game_win32_dr), but the thought came across my mind to open a discussion here.

**Comment by amerkoleci** [comment]
Yeah, definitely, feel free to share it with me or submit PR if part of repository. 

Thank you

**Reply by BMarques** [reply]
Nice, I'll share once I feel good about it. Is it ok to target windows 10 as minimum? Also targeting .NET 6.0 currently.

**Comment by amerkoleci** [comment]
Yeah, look good for me!

**Comment by MarcioAB** [comment]
Great.

I'm working on a similar "sandbox" project (work in progress).
It is based on Vortice, D3D (11 & 12), Windows Forms and net7.0
The idea is 5 separated blocks that can be mixed:
D3D_base (11 & 12), MyShaders, MyDrawSets, MyModels and WinForms GUI.

One Model can use many DrawSets and one DrawSet can be used by many Models.
One Shader can be used by many DrawSets but each DrawSet has only one Shader.
DrawSet handles generic data "upload" and draw function.
Model set DrawSets (including lights) and "upload" the meshes (MyDataSets) into the DrawSets, including Wavefront obj.
Model is based on a common Geometry class that includes one camera with basic functions.
Shader is based on a common Shader class that handles Model-View-Projection and compilation.

MyShaders, MyDrawSets and MyModels have each several ready-to-use examples that can be mixed.

Comments:
- It is a work in progress but I can share it as is.
- No documentation yet and some comments are still in Portuguese.
- D3D12 ( Level_12_0 ) is not working at all.
- No plans for Raytracing due to lack of hardware.

Example of the code for a model:

    public class MyModel6 : Geometria1
    {
        static Camera1.CameraInitialParms p = new() { cam_position = 10 * Vector3.UnitZ, cam_target = Vector3.Zero };

        // drawSets
        readonly DSA_Simplex_VertexID drawSet1;

        public MyModel6() : base(p)
        {
            GeoSetup(this);

            drawSet1 = new(); // model 6
            drawSet1.UploadData(MyDataSets.DSA_Triangulo_V3()); // model 6
        }

        public override void DrawAll()
        {
            GeoDraw();
            drawSet1.Draw(); // model 6
        }
    }

**Reply by amerkoleci** [reply]
Hi, 
Just wanted to let you know I've added D3D12 triangle example:

https://github.com/amerkoleci/Vortice.Windows.Samples/tree/main/src/Direct3D12/HelloTriangle

**Reply by MarcioAB** [reply]
Great.

By the way, this D3D12 sample also works with Shader Model 5.1 (just upgrading profile "vs_5_0" to "vs_5_1").
D3D11 works up to Shader Model 5.0 only.
Looking forward, I wonder if the D3D12 sample could be upgraded to Shader Model 6.4.

**Reply by amerkoleci** [reply]
Latest version of examples compiles with DXC and Shader Model 6.4 by default.

**Reply by MarcioAB** [reply]
HelloTexture is not ready yet, right ?

**Reply by amerkoleci** [reply]
Right, still working on it.

**Comment by BMarques** [comment]
I haven't proposed anything yet as I'm waiting to see if there's anything more to be improved regarding the memory? With the 2.1.30 I still have some issues.

**Reply by amerkoleci** [reply]
What can be still improved? Suggest so I can try to improve it

**Comment by BMarques** [comment]
My UpdateColorSpace is called every time the window moves (like the template of Chuck) which is generating this:

![image](https://user-images.githubusercontent.com/5073823/188266961-1e6ae0c8-d74b-4823-9c53-2256a0cdfac9.png)

Not sure if I can do something better on my code to prevent this.

As you can see in the image, using Description from the Output is increasing memory, The fact that EnumAdapters and EnumOutputs are IEnumerable, is also increasing memory. QueryInterfaceOrNull also contributing to the increase.

My current code is here https://github.com/BMarques/LearnDirectx/blob/42ff119abc4823151370d1747b2a45a15a009c33/BasicGameLoop/DX/DeviceResources.cs#L474

The reason why this is done is to allow support of HDR when using win + alt + b as seen here: https://github.com/walbourn/directx-vs-templates/commit/a83d12dc4a3bbc325f790870c1cfc754c78caa01
